pkg/backends/newrelic/flush.go

```diff
@@ -0,0 +1,146 @@
+	Mean        float64      `json:"samples_mean,omitempty"`
+	Median      float64      `json:"samples_median,omitempty"`
+	StdDev      float64      `json:"samples_std_dev,omitempty"`
+	Sum         float64      `json:"samples_sum,omitempty"`
+	SumSquares  float64      `json:"samples_sum_squares,omitempty"`
+	Percentiles []Percentile `json:",omitempty"`
+	Tags        []Tag        `json:",omitempty"`
+}
+
+// addMetric adds a metric to the series.
+func (f *flush) addMetric(n *Client, metricType string, value float64, persecond float64, hostname string, tags gostatsd.Tags, name string, timestamp gostatsd.Nanotime) {
+	standardMetric := setDefaultMetricSet(n, f, name, metricType, value, tags, timestamp)
+	if metricType == "counter" {
+		standardMetric[n.metricPerSecond] = persecond
+	}
+
+	f.ts.Metrics = append(f.ts.Metrics, standardMetric)
+}
+
+// addMetric adds a timer metric to the series.
+func (f *flush) addTimerMetric(n *Client, metricType string, timer gostatsd.Timer, tagsKey, name string) {
+	timerMetric := setDefaultMetricSet(n, f, name, metricType, float64(timer.Count), timer.Tags, timer.Timestamp)
+
+	if !n.disabledSubtypes.Lower {
+		timerMetric[n.timerMin] = timer.Min
+	}
+	if !n.disabledSubtypes.Upper {
+		timerMetric[n.timerMax] = timer.Max
+	}
+	if !n.disabledSubtypes.Count {
+		timerMetric[n.timerCount] = float64(timer.Count)
+	}
+	if !n.disabledSubtypes.CountPerSecond {
+		timerMetric[n.metricPerSecond] = timer.PerSecond
+	}
+	if !n.disabledSubtypes.Mean {
+		timerMetric[n.timerMean] = timer.Mean
+	}
+	if !n.disabledSubtypes.Median {
+		timerMetric[n.timerMedian] = timer.Median
+	}
+	if !n.disabledSubtypes.StdDev {
+		timerMetric[n.timerStdDev] = timer.StdDev
+	}
+	if !n.disabledSubtypes.Sum {
+		timerMetric[n.timerSum] = timer.Sum
+	}
+	if !n.disabledSubtypes.SumSquares {
+		timerMetric[n.timerSumSquares] = timer.SumSquares
+	}
+	for _, pct := range timer.Percentiles {
+		timerMetric[pct.Str] = pct.Float
+	}
+	f.ts.Metrics = append(f.ts.Metrics, timerMetric)
+}
+
+func (f *flush) maybeFlush() {
+	if uint(len(f.ts.Metrics))+20 >= f.metricsPerBatch { // flush before it reaches max size and grows the slice
+		f.cb(f.ts)
+		f.ts = &timeSeries{
+			Metrics: make([]interface{}, 0, f.metricsPerBatch),
+		}
+	}
+}
+
+func (f *flush) finish() {
+	if len(f.ts.Metrics) > 0 {
+		f.cb(f.ts)
+	}
+}
+
+func setDefaultMetricSet(n *Client, f *flush, metricName, Type string, Value float64, tags gostatsd.Tags, timestamp gostatsd.Nanotime) map[string]interface{} {
+	defaultMetricSet := map[string]interface{}{}
+	defaultMetricSet["interval"] = f.flushIntervalSec
+	defaultMetricSet["timestamp"] = timestamp
+	defaultMetricSet[n.metricType] = Type
+	defaultMetricSet[n.metricName] = metricName
+	defaultMetricSet[n.metricValue] = Value
+	defaultMetricSet["event_type"] = n.eventType
+
+	defaultMetricSet["integration_version"] = integrationVersion
+	defaultMetricSet["protocol_version"] = protocolVersion
+
+	if len(tags) > 0 {
```

Comment:
This `if` is not needed

Reply:
Good catch, removed it now.

@ash2k also agree with that comment. I think it would have been nice if there was a generic http type backend, where you could plug some additional configuration options in. Then benefit from all the work applied there.

For eg. with newrelic we could send data via the infra agent http server like this PR, or opt to send directly via our insights api.  Pros and cons to each.

Sent through the agent allows automatic applying of metadata like aws tags etc. or if sent directly to the insights api would not require you to have a resource running the infra agent and sent immediately.